aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -83,6 +83,26 @@ native::SCATTER_GATHER_OP get_operator_enum(const c10::string_view reduce) {
   }
 }
 
+TORCH_META_FUNC(gather)
+(const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  const Tensor& result = maybe_get_output(0);
+  int64_t wrapped_dim = at::maybe_wrap_dim(dim, self.dim());
+
+  if (result.defined()) {
+    at::native::resize_output(result, index.sizes());
+    at::assert_no_internal_overlap(result);
+    at::assert_no_overlap(result, self);
+    at::assert_no_partial_overlap(result, index);
+  }
+  set_output(index.sizes(), self.options());
```

Comment:
After discussing with @krshrimali about the already mentioned #63837, we converged to this solution. The description of the problem can be found in the issue, but the gist of it is:

1. resizing the output may actually only increase the tensor size without new allocation
2. the resized tensor may have memory overlap with some of the inputs
3. since memory overlap checks were already done, we fail to catch this one

@ezyang @bdhirsh  What do you think?

Reply:
That's a... really interesting case. IDK, I'm happy to chalk this up as just straight UB. Was it triggering a test?

Reply:
Hmm. I'm not very keen on the direct call to `resize_output` in the meta function. But I feel like this should work (due to implementation details of how `set_output` works):

```
  bool check_result = result.defined();
  set_output(index.sizes(), self.options());
  if (check_result) {
    at::assert_no_internal_overlap(result);
    at::assert_no_overlap(result, self);
    at::assert_no_partial_overlap(result, index);
  }
```

This definitely deserves a comment saying what's going on though...

Reply:
Thanks, @ezyang for the suggestion. This definitely looks like a better solution will gently ping again once I've committed the changes.